[15.0][MIG] sale_commission_product_criteria: full migration to v15]

PolComas opened the pull request:
[15.0][MIG] sale_commission_product_criteria: migrate to Odoo 15.0

This PR adapts the Sale Commission Product Criteria module for Odoo 15.0:

• Bump manifest version to “15.0.1.0.0” and update “depends” to ["commission", "sale_commission", "web_domain_field"].  
• Register tests in manifest (tests directory).  
• Update model inheritances:
  – _inherit “sale.commission” → “commission”  
  – _inherit “sale.commission.line.mixin” → “commission.line.mixin”  
• Fix QWeb report:
  – inherit_id “sale_commission.report_settlement” → “commission.report_settlement_document”  
  – adjust XPath to target `<table name="invoice_line_table">`.  
• Update views:
  – sale.commission form: inherit_id and model changed to “commission”/“commission.view_commission_form”  
  – sale.order.line.agent tree: clear obsolete domain on commission_id; remove settlement_type filter; hide original commission_id; insert new applied_* fields.  
• Fix demo data: model in demo XML changed from “sale.commission” → “commission”.  

All pre‑commit hooks pass, tests cover new logic, and manual validation on Odoo 15 has been performed.  
Please review and merge.

PolComas committed:
[15.0][MIG] sale_commission_product_criteria: full migration to v15]

PolComas committed:
[15.0][MIG] sale_commission_product_criteria: apply pre‑commit formatting and lint fixes

PolComas committed:
[15.0][MIG] sale_commission_product_criteria: fix pylint‑odoo warnings (newline, returns, redundant strings, context)

PolComas committed:
[15.0][MIG] sale_commission_product_criteria: remove _invoice_sale_order helper, inline invoice creation and fix pylint-odoo warnings.

PolComas committed:
[15.0][MIG] sale_commission_product_criteria: simplify tests to minimal smoke suite]

PolComas committed:
[15.0][MIG] sale_commission_product_criteria: fixed errors sale_commission_line_mixin.py

PolComas committed:
[15.0][IMP] sale_commission_product_criteria:  fixed report_settlement_template.xml

PolComas committed:
[15.0][IMP] sale_commission_product_criteria: fixed account_move.py

PolComas committed:
[15.0][MIG] sale_commission_product_criteria: implify _compute_amount for product commissions to match OCA v16 behavior and fix refund balance]

PolComas committed:
[15.0][MIG] sale_commission_product_criteria: enforce applied_on constraints in CommissionItem onchanges

ivilata reviewed (approved):
Looks good to me, thanks for your contribution!

(If any, I'd indicate in the PR title that the code comes from 14.0.)

etobella reviewed (changes requested):
Follow migration guidelines:

https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-15.0

You have lost all history of commits. Also, the migration should have just one commit.

CarlosRoca13 commented on sale_commission_product_criteria/models/commission.py:
```diff
@@ -0,0 +1,270 @@
+                    )
+                )
+
+    @api.onchange("product_id")
+    def _onchange_product_id(self):
+        if any(item.applied_on != "0_product_variant" for item in self):
+            raise ValidationError(
+                _("No es pot seleccionar un variant quan l'ítem s'aplica a '%s'.")
+                % dict(self._fields["applied_on"].selection)[self.applied_on]
+            )
+
+        for item in self:
+            item.product_tmpl_id = item.product_id.product_tmpl_id
+
+    @api.onchange("product_tmpl_id")
+    def _onchange_product_tmpl_id(self):
+        if any(item.applied_on != "1_product" for item in self):
+            raise ValidationError(
+                _("No es pot seleccionar un template quan l'ítem s'aplica a '%s'.")
+                % dict(self._fields["applied_on"].selection)[self.applied_on]
+            )
+
+        for item in self:
+            if (
+                item.product_id
+                and item.product_id.product_tmpl_id != item.product_tmpl_id
+            ):
+                item.product_id = False
```
```suggestion
    @api.onchange("product_id")
    def _onchange_product_id(self):
        has_product_id = self.filtered("product_id")
        for item in has_product_id:
            item.product_tmpl_id = item.product_id.product_tmpl_id
        if self.env.context.get("default_applied_on", False) == "1_product":
            # If a product variant is specified, apply on variants instead
            # Reset if product variant is removed
            has_product_id.update({"applied_on": "0_product_variant"})
            (self - has_product_id).update({"applied_on": "1_product"})

    @api.onchange("product_tmpl_id")
    def _onchange_product_tmpl_id(self):
        has_tmpl_id = self.filtered("product_tmpl_id")
        for item in has_tmpl_id:
            if (
                item.product_id
                and item.product_id.product_tmpl_id != item.product_tmpl_id
            ):
                item.product_id = None
```
These changes make it impossible to add products to a commission item once it has been created. Please revert it to how it was before, and if necessary (though I don't think it is), adjust it so that it continues to work.

CarlosRoca13 reviewed (changes requested):
Retrieve the commit history as indicated by @etobella, keeping only a single migration commit. Be careful not to use languages other than English in the code, and please review the comment.

PolComas commented:
Hi @etobella and @CarlosRoca13, thank you both for your detailed reviews! I’ve now:
- Pulled in the full 14.0 commit history and squashed everything into a single 15.0 migration commit per the OCA guidelines.
- Reverted the onchange methods to their original logic, translated all messages into English, and confirmed that the behavior remains identical to before.

The new PR is: https://github.com/OCA/commission/pull/631
